kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/BackoffStrategy.java

```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.filter.encryption;
+
+import java.time.Duration;
+
+public interface BackoffStrategy {
+
+    /**
+     * Decides how long to delay the next attempt at an action given it has failed
+     * for N consecutive times in the past.
+     * @param failures count of failures
+     * @return how long to delay
+     */
+    Duration getDelay(int failures);
```

Comment:
Do you foresee a backoff strategy needing to communicate "give up"?

Reply:
I don't think the BackoffStrategy should be what makes that decision, I imagine something else knowing about retry limits and combining that with the backoff strategy.

I think the Retry logic could be more involved with `CompletionStages` and trigger the `exceptionallyCompose` type logic. So it could look something more like:

```
Retrier retrier = Retrier.create(maxAttempts, executor, backoffStrategy)
..
Supplier<CompletionStage<X>> heavyWork;
CompletionStage<X> = retrier.apply(heavyWork)
```

I'm trying not to pull in the world of dependencies but I think the API of [resilience4j](https://resilience4j.readme.io/docs/retry) is something to imitate here.

Reply:
On that, I'll come back in another PR and see how it looks with a resilience lib dropped in, can debate if we want to add a dependency there.